ksonnet-gen/kubeversion/version.go

```diff
@@ -74,3 +87,97 @@ func newPropertySet(strings ...string) propertySet {
 
 	return ps
 }
+
+//-----------------------------------------------------------------------------
+// Public Data structures for specifying custom constructors for API
+// objects.
+//-----------------------------------------------------------------------------
+
+// CustomConstructorSpec specifies a custom constructor for
+// `ksonnet-gen` to emit as part of ksonnet-lib. In particular, this
+// specifies a constructor of the form:
+//
+//   foo(bar, baz):: self.bar(bar) + self.baz(baz)
+//
+// The parameter list and the body are all generated from the `Params`
+// field.
+//
+// DESIGN NOTES:
+//
+// * If the user specifies a custom constructor, we will not emit the
+//   default zero-argument constructor, `new()`. This is a purposeful
+//   decision which we make because we are typically customizing the
+//   constructors precisely because the zero-argument constructor is
+//   not meaninful for a given API object.
+// * We currently do not check that parameter names are unique.
+//   Duplicate identifiers in a parameter list results in a Jsonnet
+//   compiler error, though, so this should be caught by review and
+//   CI, and it is hence not important for this case to be covered by
+//   this code.
+type CustomConstructorSpec struct {
+	ID     string
+	Params []CustomConstructorParam
+}
+
+func newConstructor(
```

Comment:
warning extreme nit

this whitespace is inconsistent with the other constructors down below